clients/spark/examples/src/main/scala/io/treeverse/clients/examples/S3AUser.scala

```diff
@@ -0,0 +1,222 @@
+    conf.set("fs.s3a.custom.signers", "AWS4SignerType")
+
+    val region = System.getenv("AWS_REGION")
+    if (region != null) {
+      conf.set("fs.s3a.region", region)
+      conf.set("fs.s3a.endpoint", s"s3.${region}.amazonaws.com") // Otherwise it tries host-based addressing and fails
+    }
+
+    Try(new URI(args(0))) match {
+      case Failure(e) => {
+        Console.err.printf("parse URI: %s", e)
+        System.exit(1)
+      }
+      case Success(baseURI) => {
+        if (baseURI.getScheme != "s3") {
+          Console.err.printf("got scheme %s but can only handle s3\n", baseURI.getScheme)
+          System.exit(1)
+        }
+        val bucket = baseURI.getHost
+        val prefix = baseURI.getPath
+
+        def failDownload(key: String, stores: Store*): String =
+          stores.map(store =>
+            Try(store.download(key)) match {
+              case Failure(_: java.io.FileNotFoundException | _: java.util.NoSuchElementException) => null
+              case Failure(e) => s"unexpected exception ${e}"
+              case Success(contents) => s"value ${contents}, expected deleted"
+            })
+            .filter(_ != null)
+            .mkString(", ")
+
+        def checkObjectStore(store: KeyStore, objects: ObjectStore, prefix: String): String = {
+          val expected = store.list()
+          val actual = objects.list(prefix).toList
+          val tooMany = actual diff expected
+          val tooFew = expected diff actual
+          Seq(
+            (if (tooMany.isEmpty) null else  s"""unexpected objects ${tooMany.mkString(", ")}"""),
+            (if (tooFew.isEmpty) null else s"""missing objects ${tooFew.mkString(", ")}""")
+          ).filter(_ != null).mkString("; ")
+        }
+
+        def isNonemptyString(s: Any) = s match { case s: String => s != ""; case _ => false}
+
+        val s3aFS = makeS3A(conf, bucket)
+        val fs = new UnderlyingFS(s3aFS, bucket, prefix)
+        val mem = new KeyStore
+        val compare = new CompareStore(fs, mem)
+
+        val s3 = S3Client.builder().build()
+        val objects = new ObjectStore(s3, bucket)
+
+        val actions = Seq(
+          () => compare.upload("/abc", "quick brown fox"),
+          () => compare.upload("/d/e/f/xyz", "foo bar"),
+          () => checkObjectStore(mem, objects, prefix),
+          () => compare.upload("/abc/g", "slow white lemming"),
+          () => compare.download("/abc"),
+          () => compare.download("/d/e/f/xyz"),
+          () => compare.download("/abc/g"),
+          () => compare.delete("/abc"),
+          () => failDownload("/abc", fs, mem),
+          () => compare.download("/abc/g"),
+          () => checkObjectStore(mem, objects, prefix)
+        )
+
+        // TODO(ariels): Compare current state on ObjectStore after each action.
+        actions.map(_()).zipWithIndex.filter({ case (res, _) => isNonemptyString(res) }).foreach({ case (err, idx) => {
+          Console.err.println(s"[${idx}] ${err}\n")
```

Comment:
why '\n' when using `println` we already got new line

Reply:
OK :-)